phd-tests/tests/src/boot_order.rs

```diff
@@ -0,0 +1,673 @@
+        // `device_path_cursor`. To callers, this keeps it as if we had just been reading `bytes`
+        // all along.
+        bytes.set_position(bytes.position() + device_path_cursor.position());
+
+        Ok(EfiLoadOption { description, path: load_path })
+    }
+
+    fn pci_device_function(&self) -> (u8, u8) {
+        let EfiLoadPath::Device {
+            pci_device: DevicePath::Pci { device, function },
+            ..
+        } = self.path
+        else {
+            panic!(
+                "expected load path to be an ACPI/PCI pair, but was {:?}",
+                self.path
+            );
+        };
+        (device, function)
+    }
+}
+
+fn unhex(s: &str) -> Vec<u8> {
+    let s = s.replace("\n", "");
+    let mut res = Vec::new();
+    for chunk in s.as_bytes().chunks(2) {
+        assert_eq!(chunk.len(), 2);
+
+        let s = std::str::from_utf8(chunk).expect("valid string");
+
+        let b = u8::from_str_radix(s, 16).expect("can parse");
+
+        res.push(b);
+    }
+    return res;
+}
+
+async fn run_long_command(
+    vm: &phd_framework::TestVm,
+    cmd: &str,
+) -> Result<String, Error> {
+    // Ok, this is a bit whacky: something about the line wrapping for long commands causes
+    // `run_shell_command` to hang instead of ever actually seeing a response prompt.
+    //
+    // I haven't gone and debugged that; instead, chunk the input command up into segments short
+    // enough to not wrap when input, put them all in a file, then run the file.
+    vm.run_shell_command("rm cmd").await?;
+    let mut offset = 0;
+    // Escape any internal `\`. This isn't comprehensive escaping (doesn't handle \n, for
+    // example)..
+    let cmd = cmd.replace("\\", "\\\\");
+    while offset < cmd.len() {
+        let lim = std::cmp::min(cmd.len() - offset, 50);
+        let chunk = &cmd[offset..][..lim];
+        offset += lim;
+
+        // Catch this before it causes weird issues in half-executed commands.
+        //
+        // Could escape these here, but right now that's not really necessary.
+        assert!(!chunk.contains("\n"));
+
+        vm.run_shell_command(&format!("echo -n \'{}\' >>cmd", chunk)).await?;
+    }
+    vm.run_shell_command(". cmd").await
+}
+
+/// Read the EFI variable `varname` from inside the VM, and return the data therein as a byte
+/// array.
+async fn read_efivar(
+    vm: &phd_framework::TestVm,
+    varname: &str,
+) -> Result<Vec<u8>, Error> {
+    // Linux's `efivarfs` prepends 4 bytes of attributes to EFI variables.
+    let cmd = format!(
+        "dd status=none if={} bs=1 skip=4 | xxd -p -",
+        efipath(varname)
+    );
+
+    let hex = run_long_command(vm, &cmd).await?;
+
+    Ok(unhex(&hex))
+}
+
+/// Write the provided bytes to the EFI variable `varname`.
+///
+/// For Linux guests: variables automatically have their prior attributes prepended. Provide only
+/// the variable's data.
+async fn write_efivar(
```

Comment:
boring but delayed realization: writing to NvVars only actually matters if the guest OS has an EFI partition to persist to. Alpine virt images, for example, do not, so these changes are always lost on reboot. The "guest can reconfigure boot order" test will fail because of this, and rightfully so because the guest image simply does not support reconfiguring the boot order via EFI.

i think that's all basically fine, since we'd expect Alpine to basically be fine for other functionality, but now i'm curious what CI thinks of this..

Reply:
FWIW there's a `guest_os_has_read_only_fs` helper method on `TestVm` that's meant to help with cases like this--see `crucible::smoke::shutdown_persistence_test` for an example. (Of course, in that case you've just ended up not running the test, which makes CI pass but doesn't add much else of value.)

Reply:
that'd need a related `fs_is_read_readonly()` helper or something, since EFI variables are on a different filesystem, which would be fine.. but the Alpine image that doesn't seem to persist NvVars _also_ has `/sys/firmware/efi/efivars` mounted rw. i'm not confident enough to assert this yet, but i suspect somewhere in `efivarfs` the actual persistence to disk is silently lost..?

Reply:
lol: https://github.com/oxidecomputer/propolis/pull/756/commits/bcb682450ef362a98b5cb00025b392057dfaa447#diff-3a20a9ed692f02614280c3438c2e7368597cb587487dcc28be5c6bac8affc549R321-R340

the Alpine ISO 9660-backed virt image, for reference, looks like this:
```
localhost:~# mount | grep efivarfs 
efivarfs on /sys/firmware/efi/efivars type efivarfs (rw,nosuid,nodev,noexec,relatime)
```
note it even says `rw`!! i think this is because technically the filesystem could be reasonably written to even without a writable ESP - it _could_ be configuring real non-volatile storage. just, it's not here. it's unfortunate the write gets lost as a result, but it is what it is.